### PR TITLE
(SIMP-4807) Add UEFI PXEboot to tftpboot

### PR DESCRIPTION
**Pull request by lnemsick-simp**
- Added support for UEFI PXEboot.
  - Added tftpboot::linux_model_efi.
    Added tftpboot::assign_host_efi.
  - Tested primarily on CentOS 7 (grub2). Testing of CentOS 6 UEFI
    boot with legacy grub is **pending**.
- **Moved** the tftproot root directory from /tftpboot to /var/lib/tftpboot
  to satisfy the STIG RHEL-07-040520 check.
- Added tftpboot::tftpboot_root_dir to allow users to override
  the new default root directory (e.g., to set it back to /tftpboot).
- Added tftpboot::linux_install_dir and tftpboot::package_ensure
  to allow more fine-grained control of this module.
- Explicitly set selinux context on tfptboot files/directories to
  tftpdir_t in order to restrict their access to read-only tftpd
  operation.
- Internally, refactored configuration to use small, private classes.

SIMP-4807 #close

### CODE REVIEW

File: functions/install_boot_files.pp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    Hash  $os_file_info,
+    String $boot_install_dir,
+    String $package_ensure
+) {
+  # Ensure all the OS packages containing the needed boot files
+  # are installed
+  $os_packages = keys($os_file_info)
+  $os_packages.each |String $pkg_name| {
+    unless defined(Package[$pkg_name]) {
+      package { $pkg_name: ensure => $package_ensure }
```
Comment (jeefberkey) [inline functions/install_boot_files.pp]: Why create resources in a function?

File: manifests/config.pp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    mode    => '0640',
+    seltype => 'tftpdir_t'
+  }
+
+  contain 'tftpboot::config::bios'
+  contain 'tftpboot::config::efi'
+
+  if $::tftpboot::use_os_files {
+    $_os_files = tftpboot::get_os_base_filenames($::tftpboot::os_file_info)
+    $_rsync_exclude = concat([ 'pxelinux.cfg' ], $_os_files)
```
Comment (jeefberkey) [inline manifests/config.pp]: You can simply `+` these arrays together

File: manifests/config.pp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  } else {
+    $_rsync_exclude = [ 'pxelinux.cfg' ]
+  }
+
+  $_downcase_osname = downcase($facts['os']['name'])
+
+  rsync { 'tftpboot':
```
Comment (jeefberkey) [inline manifests/config.pp]: Can you put this in a boolean?

File: manifests/assign_host.pp [hunk 4]
```diff
@@ -1,55 +1,77 @@
...
+  include 'tftpboot'
 
-  file { "/tftpboot/linux-install/pxelinux.cfg/${_downname}":
-    ensure => $_ensure,
-    owner  => 'root',
-    group  => 'nobody',
-    mode   => '0644',
-    target => "templates/${model}",
-    force  => true
+  if ! ($name =~ /^\S+$/) {
```
Comment (trevor-vaughan) [inline manifests/assign_host.pp]: Nit: Should probably change this to `=~ Pattern[...]`

**Pull request merged**